File: bot_engine/notify/slack.py
"""Slack notification backend using slack-sdk."""

from __future__ import annotations

import logging

from bot_engine.notify.base import NotifyBackend

log = logging.getLogger(__name__)


class SlackNotifyBackend(NotifyBackend):
    """Sends notifications via Slack Web API."""

    def __init__(self, token: str, default_channel: str = "") -> None:
        from slack_sdk import WebClient  # lazy import — slack-sdk is optional

        self._client = WebClient(token=token)
        self._default_channel = default_channel

    def send_message(
        self,
        channel: str,
        text: str,
        *,
        thread_ts: str = "",
        unfurl_links: bool = False,
    ) -> str:
        ch = channel or self._default_channel
        if not ch:
            log.warning("No channel specified for Slack notification")
            return ""
        try:
            kwargs: dict = {
                "channel": ch,
                "text": text,
                "unfurl_links": unfurl_links,
            }
            if thread_ts:
                kwargs["thread_ts"] = thread_ts
            resp = self._client.chat_postMessage(**kwargs)
            return resp.get("ts", "") if resp.get("ok") else ""
        except Exception as e:
            log.warning("Slack notification failed (non-blocking): %s", e)
            return ""

    def add_reaction(self, channel: str, timestamp: str, emoji: str) -> None:
        ch = channel or self._default_channel
        try:
            self._client.reactions_add(channel=ch, name=emoji, timestamp=timestamp)
        except Exception:
            pass  # best-effort

File: go/internal/notify/slack.go
package notify

import (
	"context"
	"log/slog"

	"github.com/slack-go/slack"
)

// SlackNotifier sends notifications via the Slack Web API.
type SlackNotifier struct {
	client  *slack.Client
	channel string // default channel
}

// NewSlackNotifier creates a Slack-backed notifier.
// If token is empty, callers should use LogNotifier instead.
func NewSlackNotifier(token, defaultChannel string) *SlackNotifier {
	return &SlackNotifier{
		client:  slack.New(token),
		channel: defaultChannel,
	}
}

func (s *SlackNotifier) SendMessage(ctx context.Context, channel, text, threadTS string) (string, error) {
	ch := channel
	if ch == "" {
		ch = s.channel
	}
	if ch == "" {
		slog.Warn("notify/slack: no channel specified, skipping")
		return "", nil
	}

	opts := []slack.MsgOption{
		slack.MsgOptionText(text, false),
		slack.MsgOptionDisableLinkUnfurl(),
	}
	if threadTS != "" {
		opts = append(opts, slack.MsgOptionTS(threadTS))
	}

	_, ts, err := s.client.PostMessageContext(ctx, ch, opts...)
	if err != nil {
		slog.Warn("notify/slack: post message failed", "error", err)
		return "", err
	}
	return ts, nil
}

func (s *SlackNotifier) AddReaction(ctx context.Context, channel, timestamp, emoji string) error {
	ch := channel
	if ch == "" {
		ch = s.channel
	}
	err := s.client.AddReactionContext(ctx, emoji, slack.ItemRef{
		Channel:   ch,
		Timestamp: timestamp,
	})
	if err != nil {
		slog.Debug("notify/slack: add reaction failed", "error", err)
	}
	return err
}
